V 3/ Add wui-flex component

svenvoskamp opened the pull request:
# Changes

- feat: Add `wui-flex` component + `wui-spacing` css variables

0xAsimetriq commented on packages/ui/src/layout/wui-flex/index.ts:
```diff
@@ -0,0 +1,68 @@
+  @property() public padding?: string
+
+  public render() {
+    const inlineStyles = {
```
Too many options here, flex component should only be responsible for layout / flex functionality, so we need to get rid of: `backgroundColor`, `borderRadius`, `padding`, `width`, `height`

0xAsimetriq commented on packages/ui/src/layout/wui-flex/index.ts:
```diff
@@ -0,0 +1,68 @@
+      height: this.height
+    }
+
+    return html`<div style="${styleMap(inlineStyles)}"><slot></slot></div>`
```
I think we can get rid of div here and style `:host` directly in css. You should be able to do `this.styles = styleMap(inlineStyles)` inside render, to apply it directly to host as well. If not we can keep the div 👍

svenvoskamp replied:
Couldn't get this to work, so I kept the div.

0xAsimetriq commented on packages/ui/src/layout/wui-flex/index.ts:
```diff
@@ -0,0 +1,68 @@
+
+  @property() public borderRadius?: BorderRadius
+
+  @property() public flexDirection?: string
```
Instead of these being just strings, any chance we can hook into actual style types for each, to get correct auto-complete? If not, can we define them in type util and use here.

I saw you did something like this in preset util, so maybe types can be derived from there?

0xAsimetriq commented on apps/gallery/stories/layout/wui-flex.stories.ts:
```diff
@@ -0,0 +1,92 @@
+      backgroundColor="fg-300"
+      borderRadius="m"
+    >
+      <wui-flex width="40px" height="40px" backgroundColor="blue-100" borderRadius="3xl"></wui-flex>
```
Let's create a `components` folder in "gallery", for some "dumb" ones like a div which can be a different color / size and use those for preview purposes.

Also, for layout preview, maybe we can make them square (displays layout edges better than circle)

svenvoskamp commented:
@0xAsimetriq 

* Added test component
* Removed `backgroundColor`, `borderRadius`, `padding`, `width`, `height` from `wui-flex`
* Added types to properties of `wui-flex`

Could you review again?

0xAsimetriq commented on apps/gallery/components/wui-gallery-div/index.ts:
```diff
@@ -0,0 +1,32 @@
+import { html, LitElement } from 'lit'
```
To avoid confusion with ui package, let's not use `wui-` prefix, nor any imports from web3modal/ui. Storybooks components should be their own, completely detached instances. Similarly how `laboratory` will have its own things as well.

Better name like `gallery-placeholder` could be more useful.

0xAsimetriq commented on apps/gallery/components/wui-gallery-div/styles.ts:
```diff
@@ -0,0 +1,33 @@
+import { css } from 'lit'
```
Same as above, no reference and completely decouple from wui

0xAsimetriq commented on packages/ui/src/utils/TypesUtil.ts:
```diff
@@ -10,3 +10,27 @@ export type Color =
+
+export type BorderRadius = Exclude<Spacing, 'xl' | 'xxl'>
+
+export type FlexDirection = 'column-reverse' | 'column' | 'row-reverse' | 'row'
```
We can improve all of these and avoid discrepancies with PresetsUtil by doing something like

```
export type FlexWrap = typeof flexWrapOptions[number]
```

P.S for this to work you might need to do something like `flexWrapOptions as const`

Here is related stackoverflow post https://stackoverflow.com/a/54061487/911930

0xAsimetriq commented on packages/ui/src/layout/wui-flex/styles.ts:
```diff
@@ -0,0 +1,7 @@
+
+export default css`
+  div {
+    display: flex;
```
As we are keeping div in, I think we need to add `width: inherit` and `height: inherit` here, so that we can add styles on `wui-flex` from other places.

0xAsimetriq reviewed (approved):
Amazing 🔥
